File: bilibili_api/utils/varint.py
"""
bilibili_api.utils.varint

变长数字字节相关。
"""

from typing import Tuple


def read_varint(stream: bytes) -> Tuple[int, int]:
    """
    读取 varint。

    Args:
        stream (bytes): 字节流。

    Returns:
        Tuple[int, int]，真实值和占用长度。
    """
    value = 0
    position = 0
    shift = 0
    while True:
        if position >= len(stream):
            break
        byte = stream[position]
        value += (byte & 0b01111111) << shift
        if byte & 0b10000000 == 0:
            break
        position += 1
        shift += 7
    return value, position + 1

File: bilibili-api-ts/utils/varint.js
"use strict";
exports.__esModule = true;
exports.read_varint = void 0;
function read_varint(_a) {
    var stream = _a.stream;
    var value = 0;
    var position = 0;
    var shift = 0;
    while (true) {
        if (position >= stream.length) {
            break;
        }
        var byte = stream[position];
        value += (byte & 127) << shift;
        if ((byte & 128) === 0) {
            break;
        }
        position += 1;
        shift += 7;
    }
    return [value, position + 1];
}
exports.read_varint = read_varint;
